### ISSUE TITLE
The usage of collect and collectors 

### ISSUE THREAD

**Issue by Devvynneo** [nus-cs2030/2122-s1#190]
how does collect and collectors work? 
collect takes in a collector interface's method?

**Comment by FYimu** [comment]
Collectors is a class that implements the Collector interface to conduct reduction operations. And the two static methods of which we have witnessed the usefulness and application are `toList()` and `toSet()` in Lab9. Basically these two we have seen are used to collect all stream elements into a Set / List instance.

Examples:
```
Stream st = Stream.<Integer>of(1, 2, 3); // create a stream of integers 1, 2, 3
st.collect(Collectors.toSet()); // return a set of integers ==> [1, 2, 3] 
```

```
Stream st = Stream.<Integer>of(1, 2, 3); // create a stream of integers 1, 2, 3
st.collect(Collectors.toList()); // return a list of integers ==> [1, 2, 3]
```

As for the second part, "a collector interface's method" might not be so accurate as java API shows that `collect` method of Stream takes in an instance of the Collector interface. And in fact, for example, the `toSet()` and `toList()` methods in Collectors actually return object of type Collector.

What I found interesting while writing this response is that the static methods in Collectors that are used to collect objects from a stream and store them into collection do not work for primitives streams such as IntStream, and compilation error would be incurred. Take IntStream as an example, we can use `box()` to get a stream of wrapper objects or `mapToObj()`  and apply `collect()` afterwards.

Hope this helps u (it has helped myself to consolidate knowledge while searching for these haha)

**Comment by AHerdOfHLMilk** [comment]
## Always read the documentation!!!

If you are unclear, you should always refer to the documentation. 

#### But anyway

For ```stream.collect()```
You can see that there are 2 version in the documentation
one is `R collect(Supplier<R> supplier, BiConsumer<R,? super T> accumulator, BiConsumer<R,R> combiner)`
and the other is `R collect(Collector<? super T,A,R> collector)`
From the given sample code for the first one, you can roughly infer what .collect() is doing

```java
R result = supplier.get();
     for (T element : this stream)
         accumulator.accept(result, element);
     return result;
```

In fact it also most resembles `reduce` except that, the renew "reduced" result is no longer of type T but of type R.

```java
// <U> U reduce​(U identity, BiFunction<U,​? super T,​U> accumulator, BinaryOperator<U> combiner)
U result = identity;
for (T element : this stream)
    result = accumulator.apply(result, element)
return result;

// Optional<T> reduce​(BinaryOperator<T> accumulator)
boolean foundAny = false;
T result = null;
for (T element : this stream) {
    if (!foundAny) {
        foundAny = true;
        result = element;
    }
    else
        result = accumulator.apply(result, element);
}
return foundAny ? Optional.of(result) : Optional.empty();
```

This should give you some idea of what a Collector should do.

### Java collectors

> A `Collector` is specified by four functions that work together to accumulate entries into a mutable result container, and optionally perform a final transform on the result. They are:
>
> - creation of a new result container ([`supplier()`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/stream/Collector.html#supplier()))
> - incorporating a new data element into a result container ([`accumulator()`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/stream/Collector.html#accumulator()))
> - combining two result containers into one ([`combiner()`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/stream/Collector.html#combiner()))
> - performing an optional final transform on the container ([`finisher()`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/stream/Collector.html#finisher()))

tldr, 

- a collector will contain method to create some kind of containers that contains items.  
- it will have ways to add a new item into the container
- merging of 2 containers 
- some additional action on the final container

an example, say if you want to merge a stream of Strings into single String, you could use the collect method that takes in the 4 required methods directly like so:

```java
Stream<String> strings = Stream.<String>of("a","b","c");
StringBuilder result = strings.collect(StringBuilder::new, (x, y) -> x.append(y),(a, b) -> a.append(b));
System.out.println(result.toString())
```

The result will be: "abc"


Or you could use the collect method that takes in a collector and write a custom collector like so:

```java
StringCollector.java:

import java.util.Set;
import java.util.function.BiConsumer;
import java.util.function.BinaryOperator;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Collector;

class StringCollector implements Collector<String, StringBuilder, StringBuilder> {

    public static StringCollector JoinString() {
        return new StringCollector();
    }

    @Override
    public Supplier<StringBuilder> supplier() {
        return StringBuilder::new;
    }

    @Override
    public BiConsumer<StringBuilder,String> accumulator() {
        return (x, y) -> x.append(y);
    }

    @Override
    public BinaryOperator<StringBuilder> combiner() {
        return (a,b) -> a.append(b);
    }

    @Override
    public Function<StringBuilder,StringBuilder> finisher() {
        return StringBuilder::new;
    }

    @Override
    public Set<Characteristics> characteristics() {
        return Set.of(Characteristics.UNORDERED);
    }
}
```
After which, you can run:

```java
Stream<String> strings = Stream.<String>of("a","b","c");
StringBuilder result = strings.collect(StringCollector.JoinString());
System.out.println(result.toString())
```

The result will be: "abc"

Note though that creating custom collectors is probably unnecessary most of the times. just FYI.
